rust/src/info.rs

```diff
@@ -0,0 +1,287 @@
+
+    let mut codes = UInt32Builder::with_capacity(info_iter.size_hint().0);
+
+    // Type id tells which array the value is in
+    let mut type_id = UInt8BufferBuilder::new(info_iter.size_hint().0);
+    // Value offset tells the offset of the value in the respective array
+    let mut value_offsets = UInt32BufferBuilder::new(info_iter.size_hint().0);
+
+    // Make one builder per child of union array. Will combine after.
+    let mut string_values = StringBuilder::new();
+    let mut bool_values = BooleanBuilder::new();
+    let mut int64_values = Int64Builder::new();
+    let mut int32_bitmasks = Int32Builder::new();
+    let mut string_lists = ListBuilder::new(StringBuilder::new());
+    let mut int32_to_int32_list_maps = MapBuilder::new(
+        None,
+        Int32Builder::new(),
+        ListBuilder::new(Int32Builder::new()),
+    );
+
+    for (code, info) in info_iter {
+        codes.append_value(code);
+
+        match info {
+            InfoData::StringValue(val) => {
+                string_values.append_value(val);
+                type_id.append(0);
+                let value_offset = string_values.len() - 1;
+                value_offsets.append(
+                    value_offset
+                        .try_into()
+                        .expect("Array has more values than can be indexed by u32"),
+                );
+            }
+            _ => {
+                todo!("support other types in info_data")
+            }
+        };
+    }
+
+    let arrays: Vec<ArrayRef> = vec![
+        Arc::new(string_values.finish()),
+        Arc::new(bool_values.finish()),
+        Arc::new(int64_values.finish()),
+        Arc::new(int32_bitmasks.finish()),
+        Arc::new(string_lists.finish()),
+        Arc::new(int32_to_int32_list_maps.finish()),
+    ];
+    let info_schema = info_schema();
+    let union_fields = {
+        match info_schema.field(1).data_type() {
+            DataType::Union(fields, _, _) => fields,
+            _ => unreachable!(),
+        }
+    };
+    let children = union_fields
+        .iter()
+        .map(|f| f.to_owned())
+        .zip(arrays.into_iter())
+        .collect();
+    let info_value = UnionArray::try_new(
+        &[0, 1, 2, 3, 4, 5],
+        type_id.finish(),
+        Some(value_offsets.finish()),
+        children,
+    )
+    .expect("Info value array is always valid.");
+
+    let batch: RecordBatch = RecordBatch::try_new(
+        Arc::new(info_schema),
+        vec![Arc::new(codes.finish()), Arc::new(info_value)],
+    )
+    .expect("Info data batch is always valid.");
+
+    let schema = batch.schema();
+    Box::new(RecordBatchIterator::new(
+        std::iter::once(batch).map(Ok),
+        schema,
+    ))
+}
+
+pub fn import_info_data(
+    reader: Box<dyn RecordBatchReader>,
+) -> Result<Vec<(u32, InfoData)>, ArrowError> {
+    let batches = reader.collect::<Result<Vec<RecordBatch>, ArrowError>>()?;
+
+    Ok(batches
+        .iter()
+        .flat_map(|batch| {
```

Comment:
That's an issue, but also the lifetimes of the arrays are a bit messy. I tried this out, but couldn't find an easy solution. I don't think it's worth optimizing much since it's going to be handling very few values.